Use separate nodeProjectDir per project

tylerbertrand opened the pull request:
<!--
_(If you are a project committer then you may remove some/all of the following template.)_

Before creating a pull request, please file an issue in the ASF Jira system for Solr:

* https://issues.apache.org/jira/projects/SOLR

For something minor (i.e. that wouldn't be worth putting in release notes), you can skip JIRA. 
To create a Jira issue, you will need to create an account there first.

The title of the PR should reference the Jira issue number in the form:

* SOLR-####: <short description of problem or changes>

SOLR must be fully capitalized. A short description helps people scanning pull requests for items they can work on.

Properly referencing the issue in the title ensures that Jira is correctly updated with code review comments and commits. -->


# Description

Sometimes, `NpmTask`s are failing with being unable to find files/commands under `npmProjectDir`. [Here's an example of such a failure](https://ge.solutions-team.gradle.com/s/edxfitmmecv7i/failure#1). The failing tasks always execute after the `nodeSetup` task for that project, so in theory, everything should be in place.

However, in this case, multiple subprojects share the same `npmProjectDir`. The way the `gradle-node-plugin` works, having multiple projects use the same `npmProjectDir` doesn't save work, and can actually cause timing issues like the failures we are seeing. This is because each `nodeSetup` task clears out the old directory before it unpacks it again. [In the Timeline view of the same Build Scan linked above](https://ge.solutions-team.gradle.com/s/edxfitmmecv7i/timeline?details=wa5cqs7lpiyvq&end=1724359674712&outcome=failed&start=1724359674660), we can see that the `:solr:solr-ref-guide:downloadAntoraCli` task (an `NpmTask`), is executing at the same time as `:solr:packaging:nodeSetup`, and `downloadAntoraCli` can't find `node` because it's been cleared out by another suproject's `nodeSetup` task.

# Solution

This PR updates each subproject to use its own individual `nodeProjectDir` to avoid these timing issues. As mentioned above, separating these out doesn't result in any extra work being done, since sharing the same `nodeProjectDir` doesn't save work in the first place.

# Tests

No new tests were added, and no existing tests were modified, but the `check` and `integrationTests` tasks consistently pass when run locally with these changes.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://github.com/apache/solr/blob/main/CONTRIBUTING.md) and my code conforms to the standards described there to the best of my ability.
- [ ] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended, not available for branches on forks living under an organisation)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [ ] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

runningcode commented on gradle/testing/defaults-tests.gradle:
```diff
@@ -119,10 +119,12 @@ allprojects {
         jvmArgs '--illegal-access=deny'
       }
 
+      def loggingConfigFileProperty = objects.fileProperty().fileValue(file("${resources}/logging.properties"));
```
`file("${resources}/logging.properties")` is already non-lazy since at this point the file is already created. there is no value in converting it back to a non-lazy type.


I would suggest the following if `resources` is located in the project directory. If it is somewhere else there is a better way to get to it.
```suggestion
      def loggingConfigFileProperty = layout.projectDirectory.file("${resources}/logging.properties");
```

tylerbertrand replied:
It's at `gradle/testing/defaults-tests/logging.properties`

runningcode commented on gradle/testing/defaults-tests.gradle:
```diff
@@ -186,16 +188,29 @@ allprojects {
   @InputFile
   @PathSensitive(PathSensitivity.RELATIVE)
-  File loggingConfigFile
+  private final RegularFileProperty loggingConfigFile
```
the `private` keyword has no value in groovy. i would remove it.

```suggestion
  RegularFileProperty loggingConfigFile
```

runningcode commented on gradle/testing/defaults-tests.gradle:
```diff
@@ -186,16 +188,29 @@ allprojects {
 
   @Internal
-  File tempDir
+  private final RegularFileProperty tempDir
```
```suggestion
  RegularFileProperty tempDir
```

runningcode commented on gradle/testing/defaults-tests.gradle:
```diff
@@ -186,16 +188,29 @@ allprojects {
     ]
   }
+
+  RegularFileProperty getLoggingConfigFile() {
```
let's remove this and the other getter.
* groovy provides default getters for all properties
* this may override the annotations set on the private getter since it has higher visibility.

runningcode commented on gradle/testing/defaults-tests.gradle:
```diff
@@ -186,16 +188,29 @@ allprojects {
-  File tempDir
+  private final RegularFileProperty tempDir
+
+  LoggingFileArgumentProvider(RegularFileProperty loggingConfigFile, RegularFileProperty tempDir) {
```
i would remove this constructor along with the `final` keywords since it just adds boilerplate/noise to the buildscripts and groovy provides these already.

Here is a good example of how to do this: https://docs.gradle.org/current/userguide/caching_java_projects.html#dealing_with_file_paths

runningcode reviewed (commented):
I left some comments but I'm wondering if this code works. 🤔 Does changing the logging file which is marked as an input actually result in the task being re-executed? I ask because I'm not sure if the private keyword in groovy has an value and I also am not sure about the conflict between the default getters that groovy provides and the one you declared without an annotation.

tylerbertrand commented:
@runningcode It does work, changing the contents of `logging.properties` still triggers re-execution. That said I definitely agree with your assessments.

* [input comparison](https://ge.solutions-team.gradle.com/c/4lplbmz57owha/fbwxloyuphycu/task-inputs?expanded=WyJ0aDN3amNpeWZhNDI0LWp2bWFyZ3VtZW50cHJvdmlkZXJzLiQxLmxvZ2dpbmdjb25maWdmaWxlIl0)
* [build 1](https://ge.solutions-team.gradle.com/s/fbwxloyuphycu/timeline?toggled=WyI0NCJd&view=by-type#44)
* [build 2](https://ge.solutions-team.gradle.com/s/4lplbmz57owha/timeline?toggled=WyIwIl0&view=by-type#0)

tylerbertrand commented:
I changed the typing to `RegularFile` and `Directory`. The docs for `RegularFile` say that it should not reference a directory, so I used the corresponding `Directory` type for that.

Also ditched the unnecessary keywords/boilerplate for the `CommandLineArgumentProvider`s

runningcode reviewed (approved):
looks good

tylerbertrand committed:
Use separate nodeProjectDir per project

Using a shared nodeProjectDir caused multiple nodeSetup tasks to step on each others' toes. The com.github.node-gradle.node plugin doesn't fully support reusing nodeProjectDir, and each nodeSetup task will clean up the existing nodeProjectDir before unpacking the node installation into it. This can result in tasks seeing issues when trying to use npm/node while a nodeSetup task from another project is cleaning it up.

use nodeProjectDir instead of rootNodeDir

mlbiscoc committed:
SOLR-17582 Stream CLUSTERSTATUS API response (#2916)

The CLUSTERSTATUS API will now stream each collection's status to the response, fetching and computing it on the fly.  To avoid a backwards compatibility concern, this won't work for wt=javabin.

malliaridis committed:
Fix renovate.json to work with version catalogs (#2936)

cpoerschke committed:
Update copyright year in NOTICE.txt file. (#2933)

psalagnac committed:
SOLR-17471: upgrade Lucene to version 9.12.1 (#2736)

HoustonPutman committed:
SOLR-17525: Add test permissions and fix docs

alessandrobenedetti committed:
SOLR-17525: wrong model store in documentation

epugh committed:
IDE suggested cleanups (#2934)

Remove unused parameters and constants.

dsmiley committed:
SolrTestCase now supports @LogLevel (#2869)

Refactored the functionality into a TestRule, and moved from STCJ4 up to SolrTestCase.
Move reinstatement of the root log level to shutdown()

epugh committed:
Merge branch 'main' into pr/2680

epugh committed:
Track change
